src/Oscoin/Storage/Block/Abstract.hs

```diff
@@ -7,37 +7,53 @@
 -}
 
 module Oscoin.Storage.Block.Abstract
-    ( BlockStore(..)
+    ( BlockStoreReader(..)
+    , BlockStore
 
-    , hoistBlockStore
-    , noValidation
+    , hoistBlockStoreReader
     , insertBlocksNaive
     , isNovelBlock
+
+    -- * Internals
+    , BlockStoreWriter(..)
+    , hoistBlockStoreWriter
     ) where
 
 import           Oscoin.Prelude
 
-import           Oscoin.Consensus.Types (Validate)
 import           Oscoin.Crypto.Blockchain (TxLookup)
 import           Oscoin.Crypto.Blockchain.Block
 import           Oscoin.Crypto.Hash (Hashed)
+import           Oscoin.Time.Chrono (NewestFirst, OldestFirst)
+
+-- The full 'BlockStore' API is the composition of its public and private parts.
```

Comment:
It would be great if we could document the model and invariants behind a block store in details. I’d suggest the following
~~~plain
Morally, 'BlockStore' is a linked list starting with the oldest block (the genesis block)
and ending at the tip. New blocks can be appended at the end with 'insertBlock' and
the last @n@ blocks can be replaced with 'switchToFork'.
~~~

Reply:
Uhm, I am a bit on the fence with this one. The `BlockStore` per se is an opaque interface to "some persistence", and it feels strange to essentially describe it in the same terms you would do for a `Blockchain`. Furthermore, is a bit of a lie that "new blocks can be appended at the end", isn't it? In reality, we always store new things "in front", so the description here has the risk of confusing the reader even more. But even the last description is misleading, as (in reality) how we store things is purely an implementation detail of the concrete store we pick.

What about we say something along the lines of: 

> Morally, the 'BlockStore' offers an opaque interface to some kind of persistent storage for blocks, with a set of operations to insert, remove and retrieve these blocks.

Reply:
I’d be ok with that. I just wanted to make the ordering clearer. Especially that `getTip` gets the latest block from `insertBlock` and `switchToFork` replaces the most recently inserted elements. But that is probably covered by the field documentation already.